packages/server-boilerplate/src/routes/Route.ts

```diff
@@ -1,36 +1,37 @@
-/*
+/**
  * Tupaia
- * Copyright (c) 2017 - 2020 Beyond Essential Systems Pty Ltd
- *
+ * Copyright (c) 2017 - 2021 Beyond Essential Systems Pty Ltd
  */
-import { NextFunction, Response } from 'express';
-import { respond, UnauthenticatedError } from '@tupaia/utils';
-
-import { TupaiaRequest, TupaiaResponseBody, SessionCookie } from '../types';
-import { SessionModel, SessionType } from '../models/Session';
-
-export abstract class Route {
-  req: TupaiaRequest;
-
-  res: Response;
-
-  next: NextFunction;
-
-  sessionModel: SessionModel;
-
-  sessionCookie?: SessionCookie;
-
-  constructor(req: TupaiaRequest, res: Response, next: NextFunction) {
+import { Request, NextFunction, Response } from 'express';
+import { respond } from '@tupaia/utils';
+
+// Infers type arguments from request type
+/* eslint-disable @typescript-eslint/no-explicit-any */
+export type Params<Type> = Type extends Request<infer P, any, any, any> ? P : never;
+export type ResBody<Type> = Type extends Request<any, infer sBody, any, any> ? sBody : never;
+export type ReqBody<Type> = Type extends Request<any, any, infer qBody, any> ? qBody : never;
+export type Query<Type> = Type extends Request<any, any, any, infer Q> ? Q : never;
+export type MatchingRequest<Req> = Request<Params<Req>, ResBody<Req>, ReqBody<Req>, Query<Req>>;
+export type MatchingResponse<Req> = Response<ResBody<Req>>;
+/* eslint-enable @typescript-eslint/no-explicit-any */
+
+export class Route<
+  Req extends MatchingRequest<Req> = Request,
```

Comment:
My understanding is that we use `MatchingRequest`/`MatchingResponse` to match the passed in types to `express` types (although I may be wrong).

If that's indeed the case, perhaps we should mention `Express` in the types to make our intentions clearer, eg `ExpressRequest`/`ExpressResponse`

Reply:
Good point, I think that is a bit more clear 😄